In [1]:
#importing necessary libraries
import os 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings
warnings.simplefilter('ignore', FutureWarning)

In [2]:
#loading the Avila dataset
dat=pd.read_csv("Avila-DataSet for miniprojects.csv")

In [3]:
#finding the dimension of the dataset
dat.shape

(20867, 11)

In [4]:
#finding the first five rows from the dataset
dat.head()

Intercolumnar distance   upper margin    lower margin   exploitation   \
0                 0.266074      -0.165620        0.320980       0.483299   
1                 0.130292       0.870736       -3.210528       0.062493   
2                -0.116585       0.069915        0.068476      -0.783147   
3                 0.031541       0.297600       -3.210528      -0.583590   
4                 0.229043       0.807926       -0.052442       0.082634   

   row number   modular ratio   interlinear spacing    weight    peak number   \
0     0.172340        0.273364              0.371178  0.929823       0.251173   
1     0.261718        1.436060              1.465940  0.636203       0.282354   
2     0.261718        0.439463             -0.081827 -0.888236      -0.123005   
3    -0.721442       -0.307984              0.710932  1.051693       0.594169   
4     0.261718        0.148790              0.635431  0.051062       0.032902   

   modular ratio/ interlinear spacing  \
0                            0.159345   
1                            0.515587   
2                            0.582939   
3                           -0.533994   
4                           -0.086652   

  Class: A, B, C, D, E, F, G, H, I, W, X, Y  
0                                         A  
1                                         A  
2                                         A  
3                                         A  
4                                         F

In [5]:
#finding the null values
dat.isnull().sum()

Intercolumnar distance                       0
upper margin                                 0
 lower margin                                0
exploitation                                 0
row number                                   0
modular ratio                                0
interlinear spacing                          0
weight                                       0
 peak number                                 0
modular ratio/ interlinear spacing           0
Class: A, B, C, D, E, F, G, H, I, W, X, Y    0
dtype: int64

In [6]:
#replacing column name
dat["Class"]=dat["Class: A, B, C, D, E, F, G, H, I, W, X, Y"]

In [7]:
#removing the Class: A, B, C, D, E, F, G, H, I, W, X, Y column
dat.drop("Class: A, B, C, D, E, F, G, H, I, W, X, Y",axis="columns",inplace=True)

In [8]:
#selecting only A,E,F,H classes
dat_sub=dat[dat["Class"].isin(['A','E','F','H'])] 

In [9]:
#obtaining the unique elements
dat_sub.Class.unique()

array(['A', 'F', 'H', 'E'], dtype=object)

In [10]:
#importing the LabelEncoder from sklearn
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

In [11]:
#fitting the model to encode class:A=0,E=1,F=2,H=3
le.fit(dat_sub["Class"])

LabelEncoder()

In [12]:
#assigning the encoded values to the Classes A,F,H,E
dat_sub["encoded_class"]=le.transform(dat_sub["Class"])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [13]:
#printing the dataframe
dat_sub

Intercolumnar distance   upper margin    lower margin   exploitation   \
0                     0.266074      -0.165620        0.320980       0.483299   
1                     0.130292       0.870736       -3.210528       0.062493   
2                    -0.116585       0.069915        0.068476      -0.783147   
3                     0.031541       0.297600       -3.210528      -0.583590   
4                     0.229043       0.807926       -0.052442       0.082634   
...                        ...            ...             ...            ...   
20859                -0.017834       0.234790       -0.180472       0.621032   
20861                -0.437525       0.423218        0.388552       0.620852   
20864                -0.054866       0.580242        0.032912      -0.016668   
20865                 0.080916       0.588093        0.015130       0.002250   
20866                 0.377169       0.014957        0.381439       0.292753   

       row number   modular ratio   interlinear spacing    weight   \
0         0.172340        0.273364              0.371178  0.929823   
1         0.261718        1.436060              1.465940  0.636203   
2         0.261718        0.439463             -0.081827 -0.888236   
3        -0.721442       -0.307984              0.710932  1.051693   
4         0.261718        0.148790              0.635431  0.051062   
...            ...             ...                   ...       ...   
20859     0.261718        0.439463              0.673182 -0.262982   
20861     0.172340       -0.889332              0.144676  2.131034   
20864     0.261718        1.519109              0.371178 -0.985508   
20865     0.261718       -0.930856             -0.270579  0.163807   
20866     0.261718       -1.470679             -0.006326 -0.494919   

        peak number   modular ratio/ interlinear spacing Class  encoded_class  
0           0.251173                            0.159345     A              0  
1           0.282354                            0.515587     A              0  
2          -0.123005                            0.582939     A              0  
3           0.594169                           -0.533994     A              0  
4           0.032902                           -0.086652     F              2  
...              ...                                 ...   ...            ...  
20859       0.064084                            0.140600     F              2  
20861       1.373706                           -0.771516     E              1  
20864      -0.403638                            1.276301     A              0  
20865      -0.091823                           -0.593329     F              2  
20866      -0.247731                           -1.212974     H              3  

[15724 rows x 12 columns]

In [14]:
#taking X values as all the 10 columns except Class and encoded_class 
X=dat_sub.iloc[:,:10]

In [15]:
#taking y values as encoded_class
y=dat_sub["encoded_class"]

In [16]:
#importing train_test_split from sklearn library
from sklearn.model_selection import train_test_split

In [17]:
#splitting the dataset into training and testing as 80% and 20%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

In [18]:
#importing SVC(support vector classifier) from svm (support vector machine)
from sklearn.svm import SVC

In [19]:
# Instantiate the Support Vector Classifier (SVC) with defaults
svc = SVC()
 
# Fitting the model
svc.fit(X_train, y_train)

SVC()

In [20]:
#Predict the response for test dataset
y_pred = svc.predict(X_test)
y_pred

array([0, 0, 0, ..., 0, 1, 0])

In [35]:
#importing mean_squared_error(MSE) from sklearn library and finding MSE
from sklearn.metrics import mean_squared_error
mse_default=mean_squared_error(y_test,y_pred)
print('Mean Squared Error is:',mse_default)

Mean Squared Error is: 1.1440381558028616


In [32]:
pip install dmba

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 11.8 MB 4.7 MB/s 


In [34]:
#importing the classificationSummary from Data mining business analytics library
#finding the classificationSummary for both training and testing 
from dmba import classificationSummary
# training  
classificationSummary(y_train, svc.predict(X_train))
# test  
classificationSummary(y_test, svc.predict(X_test))

Confusion Matrix (Accuracy 0.7280)

       Prediction
Actual    0    1    2    3
     0 6348  118  336   55
     1  319 1295   74   52
     2 1985   59 1087   14
     3  294   49   66  428
Confusion Matrix (Accuracy 0.7135)

       Prediction
Actual    0    1    2    3
     0 1588   15  101   11
     1   89  327   24   10
     2  516   18  243    1
     3   88    8   20   86


*For SVC Default Model* 

Accuracy for training data is 72.80%


Accuracy for testing data is 71.35%


In [21]:
from sklearn import metrics
print(metrics.classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.70      0.93      0.79      1715
           1       0.89      0.73      0.80       450
           2       0.63      0.31      0.42       778
           3       0.80      0.43      0.55       202

    accuracy                           0.71      3145
   macro avg       0.75      0.60      0.64      3145
weighted avg       0.71      0.71      0.69      3145



In [26]:
#calculating balanced_accuracy_score
balanced_accuracy_score=metrics.balanced_accuracy_score(y_test,y_pred)
print("Balanced accuracy score for SVC default model = ",balanced_accuracy_score)

Balanced accuracy score for SVC default model =  0.5976740236023798


In [25]:
#calcutating balanced_error_rate (which is 1-balanced_accuracy_score)
Balanced_error_rate=1-balanced_accuracy_score
print("Balanced error rate for SVC default model = ",Balanced_error_rate)

Balanced error rate for SVC default model =  0.40232597639762024


###Trying to increase our model performance by Hyperparameter tuning using GridSearchCV

In [28]:
#improving models performance by using GridSearchCV
from sklearn.model_selection import GridSearchCV
  
# defining parameter range
param_grid = {'C': [0.1, 1, 10, 100, 1000], 
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
              'kernel': ['rbf']} 
  
grid = GridSearchCV(SVC(), param_grid, refit = True, verbose = 3)
  
# fitting the model for grid search
grid.fit(X_train, y_train)

Fitting 5 folds for each of 25 candidates, totalling 125 fits
[CV 1/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.674 total time=   7.9s
[CV 2/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.693 total time=   7.8s
[CV 3/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.689 total time=   9.2s
[CV 4/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.681 total time=   9.0s
[CV 5/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.692 total time=   9.3s
[CV 1/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.664 total time=   7.4s
[CV 2/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.669 total time=   7.4s
[CV 3/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.668 total time=   8.7s
[CV 4/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.674 total time=   6.6s
[CV 5/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.680 total time=   8.9s
[CV 1/5] END .....C=0.1, gamma=0.01, kernel=rbf;, score=0.545 total time=   9.4s
[CV 2/5] END .....C=0.1, gamma=0.01, kernel=rbf

GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.1, 1, 10, 100, 1000],
                         'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
                         'kernel': ['rbf']},
             verbose=3)

In [29]:
# printing best parameter after tuning
print(grid.best_params_)
  
# printing how our model looks after hyper-parameter tuning
print(grid.best_estimator_)

{'C': 100, 'gamma': 1, 'kernel': 'rbf'}
SVC(C=100, gamma=1)


In [30]:
#Predict the response for test dataset
grid_predictions = grid.predict(X_test)
  
# printing classification report
print(metrics.classification_report(y_test, grid_predictions))

              precision    recall  f1-score   support

           0       0.86      0.91      0.89      1715
           1       0.94      0.86      0.90       450
           2       0.80      0.76      0.78       778
           3       0.87      0.78      0.83       202

    accuracy                           0.86      3145
   macro avg       0.87      0.83      0.85      3145
weighted avg       0.86      0.86      0.86      3145



In [36]:
#finding the classificationSummary for both training and testing 
# training  
classificationSummary(y_train, grid.predict(X_train))
# test  
classificationSummary(y_test, grid.predict(X_test))

Confusion Matrix (Accuracy 0.9793)

       Prediction
Actual    0    1    2    3
     0 6747    0  108    2
     1    0 1740    0    0
     2  141    0 3003    1
     3    7    0    2  828
Confusion Matrix (Accuracy 0.8588)

       Prediction
Actual    0    1    2    3
     0 1564    7  127   17
     1   46  389   11    4
     2  177    9  590    2
     3   25    7   12  158


*For SVC Model after tuning* 

Accuracy for training data is 97.93%


Accuracy for testing data is 85.88%

In [41]:
#importing mean_squared_error(MSE) from sklearn library and finding MSE
from sklearn.metrics import mean_squared_error
mse_after_tuning=mean_squared_error(y_test,grid_predictions)
print('Mean Squared Error is:',mse_after_tuning)

Mean Squared Error is: 0.548489666136725


In [40]:
#calculating balanced_accuracy_score
balanced_accuracy_score1=metrics.balanced_accuracy_score(y_test,grid_predictions)
print("Balanced accuracy score for SVC model after tuning= ",balanced_accuracy_score1)

Balanced accuracy score for SVC model after tuning=  0.829232692704992


In [39]:
#calcutating balanced_error_rate (which is 1-balanced_accuracy_score)
Balanced_error_rate1=1-balanced_accuracy_score1
print("Balanced error rate for SVC model after tuning = ",Balanced_error_rate1)

Balanced error rate for SVC model after tuning =  0.170767307295008


As we know that lower the error better the model and higher the accuracy the better the performance of the model.
We can conclude from the above table after performing **Hyperparameter tuning** which resulted in **lower mean squared error** is the better model. And, the **Accuracy** for training and testing of the model also **increased**.


*   **Balanced_error_rate** of the model after performing tuning resulted to have **lower** than the model before tuning.






